slimt/Annotation.cc

```diff
@@ -74,4 +77,72 @@ void AnnotatedText::record_existing_sentence(
   annotation.token_begin_.push_back(text.size());
 }
 
+void AnnotatedText::to(Encoding encoding) {
+  if (encoding == Encoding::UTF8) {
+    WordIterator current(*this);
+
+    std::vector<Range> words;
+
+    size_t utf8_idx = 0;
+    size_t byte_idx = 0;
+    Range utf8{
+        //
+        .begin = utf8_idx,  //
+        .end = 0            //
+    };
+    // This loops run on the entire string.
+    //
+    // We have indices into two views of the same string. One is bytes, the
+    // other is utf8 encoded. We want to convert what is bytes to utf8.
+    //
+    // For this, we traverse through the characters, checking for unicode
+    // encoding and associated adjustments to the utf8_idx, keeping
+    // correspondences with the byte_idx;
+    for (const char c : text) {
+      // current = [begin, end)
+      if ((c & 0xc0) != 0x80)  // if is not utf-8 continuation character
+        ++utf8_idx;
+
+      ++byte_idx;
+
+      if (byte_idx == (*current).end) {
+        utf8.end = utf8_idx;
+
+        // Push it to the list of (utf8) words.
+        // We will use these to create the utf8 range annotation.
+        words.push_back(utf8);
+        ++current;
+
+        utf8.begin = utf8_idx;
+      }
+    }
+    annotation.update(words);
+  }
+}
+
+WordIterator &WordIterator::operator++() {
+  ++word_idx_;
+  if (word_idx_ >= annotated_.word_count(sentence_idx_)) {
+    ++sentence_idx_;
+    word_idx_ = 0;
+  }
+  return *this;
+}
+
+WordIterator &WordIterator::operator--() {
+  if (word_idx_ == 0) {
+    --sentence_idx_;
+    word_idx_ = annotated_.word_count(sentence_idx_) - 1;
+  } else {
+    --word_idx_;
+  }
+
+  return *this;
+}
```

Comment:
Do we use `--`? I realize I mentioned `--` to motivate iterator, but I also remember saying bare minimum required. Two things I learned from past:

1. Be careful around `--` on `size_t`. Because after 0, it overflows and stuff. 
2. `if` (or branches) inside these operators are not preferable. When we relax continuity constraints, we should be able to simply use the container's iterator `std::vector<Range>`. Right now we have to generate the Word from begin markers which prevent us from doing it.